rust/core-lib/src/view.rs

```diff
@@ -125,8 +126,48 @@ impl View {
         self.cursor_col = col;
     }
 
-    pub fn get_cursor_col(&mut self) -> usize {
-        self.cursor_col
+    /// Toggles a caret at the given offset.
+    pub fn toggle_sel(&mut self, offset: usize) {
+        self.sel_dirty = true;
+        if !self.selection.regions_in_range(offset, offset).is_empty() {
+            self.selection.delete_range(offset, offset);
+            if !self.selection.is_empty() {
+                return;
+            }
+        }
+        let region = SelRegion {
+            start: offset,
+            end: offset,
+            // TODO: might want to set horiz to some meaningful value.
+            horiz: None,
+            affinity: Affinity::default(),
+        };
+        self.selection.add_region(region);
+    }
+
+    /// Move the selection by the given movement. Return value is the offset of
+    /// a point that should be scrolled into view.
+    ///
+    /// If `modify` is `true`, the selections are modified, otherwise the results
+    /// of individual region movements become carets.
+    pub fn do_move(&mut self, text: &Rope, movement: Movement, modify: bool)
+        -> Option<usize>
+    {
+        let new_selection = selection_movement(movement, &self.selection, self, text,
+            modify);
+        self.selection = new_selection;
+        // We somewhat arbitrarily choose the last region for setting the old-style
+        // selection state, and for scrolling it into view if needed. This choice can
```

Comment:
That matches my measurements. I'm not 100% sure it's the best behavior, but it does have being symmetrical going for it.

At this stage in the game, I have to choose a single point. When `sel_start` and `sel_end` are actually deleted out of View, it will be easier to think about the best behavior.

Very likely this method will want to return a range rather than a single point. That would make it fairly easy to replicate the sublime behavior, or do something else equally reasonable.

One thing I'll just throw out there because it's simple and would be easy to implement: keeping the midpoint in view.